### ISSUE TITLE
Is it normal to leak _notify_atexit functions?

### ISSUE THREAD

**Issue by linedu** [zeromq/pyzmq#767]
libzmq-4.0.4
pyzmq-14.3.0
Python-2.7.3

I've been investigating memory leaks in an application and we run a long-running web server solution (Django-based). It performs RPC-requests using a ZeroMQ client (basically Paranoid Pirate stuff from ZeroMQ guide).

Even when we destroy a context after using it I'm seeing every ZeroMQ request generating an additional function registration that will not be picked up by the garbage collector. To my understanding because it will be referenced as an exit handler.

Is this normal?

Code to try it:
```python
# pip install zmq objgraph pympler

import zmq
import objgraph
import gc
from pympler import muppy
import types

print "Creating baseline for objgraph..."
objgraph.show_growth()
print "Ignore above."

context = zmq.Context(1)
print "Created context."
context.term()
print "Terminated context."
del context
print "Removed reference to context"
collected = gc.collect()
print "Forced garbage collection. Collected %d" % collected
print "New stuff since last check after garbage collection:"
objgraph.show_growth()

print "What is that new function?"
all_objects = muppy.get_objects()
funcs = muppy.filter(all_objects, Type=types.FunctionType)
fu = funcs[-1]
print "Probably: %s" % fu.__name__
```

**Comment by lawik** [comment]
Posted this under my work account. I'll just post here so I can follow :)

**Comment by vromanov** [comment]
The same result with Python 3.4.3.

**Comment by minrk** [comment]
I'm not sure what you mean by leak. There is a single, global garbage collector. Leak suggests that these are being created over and over again. Are you seeing that?

**Comment by lawik** [comment]
I am seeing a new _notify_atexit function being created every time a context is started and it is not cleaned up. It looks like the code is basically a factory for these functions, a new one is defined each time it is called, it rhymes with what I'm seeing at least.
So I make an RPC call using ZeroMQ. 1 function created according to objgraph
Next RPC call. 2 functions created.
Another RPC call. 3 functions created.

After heavy testing I had a lot of these functions.
From my testing and the script above I see multiple instances of cleanup functions, not a single one registered multiple times.

**Comment by minrk** [comment]
Gotcha, there's one that Contexts create that I forgot about. I'll make sure they get cleaned up.

**Comment by minrk** [comment]
Should be fixed by #769

**Comment by minrk** [comment]
A side note: From your description, it sounds like you are creating and destroying a zmq Context for each RPC call. That's probably not what you should do, regardless of this particular issue.
